### DISCUSSION TITLE
Connect flutter client

### DISCUSSION THREAD

**Discussion by fredchess** [soketi/pws#1059] [Q&A]
Hello,
Have a question. I have a laravel api and flutter app. 
So can I use socket.io client on my flutter app? If no, how can I connect flutter to my soketi server?

Thanks.

**Comment by afridishaikh07** [comment]
Yes, you can use a Socket.IO client in your Flutter app to connect to a Socket.IO server. Use https://pub.dev/packages/socket_io_client

**Comment by fsdevchivalry** [comment]
Hello fredchess,
hope this code will help you.

     // Initialize the socket connection
      void initSocket({required String username, required String token}) async {
        _client = io.io('wss://xx.xxx.com:449', <String, dynamic>{
          'transports': ['websocket'],
          'autoConnect': false,
        }); // Connect to the server
        await emitConnect(username, token);
      }

     // Async funtion
     Future<void> emitConnect(String username, String accessToken) async {
      try {
        if (_client != null) {
          _logger.debug('attempting to connect to socket');
          _client?.onConnect((args) {
            try {
              _logger.info("Socket Connected");
              _logger.debug("Socket.Id: ${_client?.id}");
              // Add all on functions here
              //seperate into functions
              _client!.on('join', (data) {
                _logger.info('Joined: $data');
                isJoined = true;
              });
  
             
              _client?.onDisconnect((error) {
                _logger.info('Disconnected from socket $error');
              });
  
              emitJoin(username, accessToken);
            } catch (ex) {
              // displayReportResultTrack(ex);
            }
          });
          _client?.onConnectError(
              (data) => {_logger.error('Failed to connect https')});
          _client?.connect();
        }
      } catch (ex) {
      }
    }

**Reply by merlinstwart0922** [reply]
That's so useful.

**Comment by fsdevchivalry** [comment]
Maybe you can visit the https://pub.dev/packages/socket_io_client.
There are all documentation about the socket.io connection with flutter.

**Comment by jonasermert** [comment]
```
dependencies:
  socket_io_client: ^2.0.0
```

```
import 'package:flutter/material.dart';
import 'package:socket_io_client/socket_io_client.dart' as io;

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  io.Socket socket;

  @override
  void initState() {
    super.initState();
    // Connect to your socket.io server
    socket = io.io('http://your_socket_server_url');
    socket.on('connect', (_) {
      print('Connected to server');
    });
    socket.on('message', (data) {
      print('Received message: $data');
    });
    socket.connect();
  }

  @override
  void dispose() {
    // Disconnect from the socket when the widget is disposed
    socket.disconnect();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Socket.io Flutter App'),
      ),
      body: Center(
        child: Text('Socket.io connection established!'),
      ),
    );
  }
}
```

**Comment by ceojazzy** [comment]
i am still having issues connection it says:

socket error: App key "my key from railwaiapp" not in this cluster. Did you forget to specify the cluster?
I am using pusher pusher_client: ^2.0.0

i deployed on railway.app and code is here:

`void connectPusher() {
    print('Connecting to Pusher...');

    // Replace these with your Pusher credentials
    String appId = 'appid from railway app';
    String key = 'key form railway app variables';
    String secret = 'secrect from railway app variablesl';

    PusherOptions options = PusherOptions(
      host: 'soketi-production-456c.up.railway.app',
      wsPort: 6001,
      encrypted: false,
    );

    pusher = PusherClient(
      key,
      PusherOptions(
        encrypted: true,
      ),
      enableLogging: true,
    );

    pusher.connect();

    pusher.onConnectionStateChange((state) {
      print(
          "socket previousState: ${state!.previousState}, currentState: ${state.currentState}");
    });

    pusher.onConnectionError((error) {
      print("socket error: ${error!.message}");
    });

// Subscribe to a private channel
    Channel channel = pusher.subscribe(widget.matchId);

// Bind to listen for events called "order-status-updated" sent to "private-orders" channel
    channel.bind(widget.matchId, (PusherEvent? event) {
      print('Socket event data? ${event!.data}');
    });

// Unsubscribe from channel
    // pusher.unsubscribe("private-orders");

// Disconnect from pusher service
    // pusher.disconnect();
  }`

**Comment by jonasermert** [comment]
The issue you're experiencing with the Pusher client could be due to the incorrect configuration of the PusherOptions, specifically the cluster and host settings. Based on your provided code, it seems like the PusherOptions configuration might be incomplete or incorrectly set. Modify your connectPusher function to include the cluster information and the host configuration:

```
void connectPusher() {
  print('Connecting to Pusher...');

  // Replace these with your Pusher credentials
  String appId = 'appid from railway app';
  String key = 'key from railway app variables';
  String secret = 'secret from railway app variables';

  PusherOptions options = PusherOptions(
    host: 'soketi-production-456c.up.railway.app',
    port: 6001,
    encrypted: false,
    cluster: 'your_cluster_name', // Replace with your Pusher cluster name
  );

  pusher = PusherClient(
    key,
    options,
    enableLogging: true,
  );

  pusher.connect();

  pusher.onConnectionStateChange((state) {
    print("socket previousState: ${state!.previousState}, currentState: ${state.currentState}");
  });

  pusher.onConnectionError((error) {
    print("socket error: ${error!.message}");
  });

  Channel channel = pusher.subscribe(widget.matchId);

  // Bind to listen for events called "order-status-updated" sent to "private-orders" channel
  channel.bind(widget.matchId, (PusherEvent? event) {
    print('Socket event data? ${event!.data}');
  });

  // Unsubscribe from channel
  // pusher.unsubscribe("private-orders");

  // Disconnect from pusher service
  // pusher.disconnect();
}

```

**Comment by ceojazzy** [comment]
Hi @jonasermert i thought we can use Soketi as an individual socket web server. if i create an account on pusher then i will fall in there pricing right? i don't want to use pusher as there prices are expensive.
